File: app/src/main/java/com/vocaby/app/data/VocabyDatabase.java
package com.vocaby.app.data;

import android.content.Context;

import androidx.room.Database;
import androidx.room.Room;
import androidx.room.RoomDatabase;

import com.vocaby.app.data.dao.VocabyDao;
import com.vocaby.app.data.entity.CustomDefinition;
import com.vocaby.app.data.entity.CustomEntry;
import com.vocaby.app.data.entity.CustomEntryGroup;
import com.vocaby.app.data.entity.Definition;
import com.vocaby.app.data.entity.Type;
import com.vocaby.app.data.entity.User;
import com.vocaby.app.data.entity.UserSave;
import com.vocaby.app.data.entity.Word;

@Database(entities = {
        Word.class, User.class, Definition.class, UserSave.class,
        CustomEntry.class, CustomEntryGroup.class, CustomDefinition.class, Type.class},
        version = 1, exportSchema = false)
public abstract class VocabyDatabase extends RoomDatabase {
    public abstract VocabyDao vocabyDao();
    private static volatile VocabyDatabase INSTANCE;

    public static VocabyDatabase getDatabase(final Context context) {
        if (INSTANCE == null) {
            synchronized (VocabyDatabase.class) {
                if (INSTANCE == null) {
                    INSTANCE = Room.databaseBuilder(context.getApplicationContext(),
                            VocabyDatabase.class, "database")
                            .createFromAsset("databases/database.db")
                            .allowMainThreadQueries()
                            .build();
                }
            }
        }

        return INSTANCE;
    }
}

File: app/src/main/java/com/vocaby/application/data/VocabyDatabase.kt
package com.vocaby.application.data

import android.content.Context
import androidx.room.AutoMigration
import androidx.room.Database
import androidx.room.Room
import androidx.room.RoomDatabase
import androidx.sqlite.db.SupportSQLiteDatabase
import com.vocaby.application.data.dao.VocabyDao
import com.vocaby.application.data.entity.*
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch

@Database(entities = [User::class, UserSave::class,
    Type::class, Word::class, Definition::class, CustomEntry::class, CustomDefinition::class,
    CustomEntryGroup::class, DictionaryViewCount::class, CustomDictionaryViewCount::class],
    version = 2,
    exportSchema = true,
    autoMigrations = [
        AutoMigration (from = 1, to = 2)
    ]
)
abstract class VocabyDatabase : RoomDatabase() {
    abstract fun vocabyDao() : VocabyDao

    companion object {
        @Volatile
        private var INSTANCE: VocabyDatabase? = null

        fun getDatabase(context: Context, scope: CoroutineScope): VocabyDatabase {
            // if the INSTANCE is not null, then return it,
            // if it is, then create the database
            return INSTANCE ?: synchronized(this) {
                val instance = Room.databaseBuilder(
                    context.applicationContext,
                    VocabyDatabase::class.java,
                    "database"
                ).createFromAsset("databases/vocabydb.db")
                .setJournalMode(JournalMode.TRUNCATE)
                .addCallback(Callback(scope))
                .build()

                INSTANCE = instance
                instance
            }
        }
    }

    private class Callback(
        private val scope: CoroutineScope
    ) : RoomDatabase.Callback() {
        override fun onCreate(db: SupportSQLiteDatabase) {
            super.onCreate(db)
            INSTANCE?.let { database ->
                scope.launch(Dispatchers.IO) {
                    populateDatabase(database.vocabyDao())
                }
            }
        }

        fun populateDatabase(dao: VocabyDao) {
            val types = listOf(
                Type("noun"),
                Type("verb"),
                Type("adjective"),
                Type("adverb"),
                Type("idiom"),
                Type("proverb"),
                Type("phrase"),
                Type("preposition"),
                Type("interjection"),
                Type("conjunction"),
                Type("pronoun"),
            ).toTypedArray()

            dao.insertTypes(*types)
        }
    }
}